lib/osut/utils.rb

```diff
@@ -7086,55 +7432,162 @@ def addSkyLights(spaces = [], opts = {})
+    # Delete incomplete sets (same as rejected if 'voided').
+    sets.reject! { |set| set.key?(:void) }
+    sets.select! { |set| set.key?(:pattern) }
+    return empty("sets (3)", mth, WRN, rm2) if sets.empty?
+
+    # Skylight size contraction if overshot (e.g. scale down by -13% if > +13%).
+    # Applied on a surface/pattern basis: individual skylight sizes may vary
+    # from one surface to the next, depending on respective patterns.
+
+    # First, skip whole sets altogether if their total m2 < (skm2 - sm2). Only
+    # considered if significant discrepancies vs average set skylight m2.
+    sbm2 = 0
+
+    sets.each do |set|
+      sbm2 += set[:cols] * set[:w] * set[:rows] * set[:d] * set[:m]
+    end
+
+    avm2 = sbm2 / sets.size
+
+    if skm2.round(2) > sm2.round(2)
+      sets.reverse.each do |set|
+        break unless skm2.round(2) > sm2.round(2)
+
+        stm2 = set[:cols] * set[:w] * set[:rows] * set[:d] * set[:m]
+        next unless stm2 < 0.75 * avm2
+        next unless stm2.round(2) < (skm2 - sm2).round(2)
+
+        skm2 -= stm2
+        set[:void] = true
+      end
+    end
+
+    sets.reject! { |set| set.key?(:void) }
+    return empty("sets (4)", mth, WRN, rm2) if sets.empty?
+
+    # Size contraction: round 1: low-hanging fruit.
     if skm2.round(2) > sm2.round(2)
       ratio2 = 1 - (skm2 - sm2) / skm2
       ratio  = Math.sqrt(ratio2)
-      skm2  *= ratio2
 
       sets.each do |set|
-        next if set.key?(:void)
-        next unless set.key?(:pattern)
+        am2 = set[:cols] * set[:w] * set[:rows] * set[:d] * set[:m]
+        xr  = set[:w]
+        yr  = set[:d]
 
-        pattern = set[:pattern]
-        next unless set.key?(pattern)
+        if xr > w0
+          xr = xr * ratio < w0 ? w0 : xr * ratio
+        end
 
-        case pattern
-        when "array" # equally adjust both width and depth
-          xr  = set[:w] * ratio
-          yr  = set[:d] * ratio
-          dyr = set[:d] - yr
-
-          set[:w ]  = xr
-          set[:d ]  = yr
-          set[:w0]  = set[:w] + gap
-          set[:d0]  = set[:d] + gap
-          set[:dY] += dyr / 2
-        when "strips" # adjust depth
-          xr2 = set[:w] * ratio2
-
-          set[:w ]  = xr2
-          set[:w0]  = set[:w] + gap
-        else # "strip", adjust width
-          yr2 = set[:d] * ratio2
-          dyr = set[:d] - yr2
-
-          set[:d ]  = yr2
-          set[:d0]  = set[:w] + gap
-          set[:dY] += dyr / 2
+        if yr > w0
+          yr = yr * ratio < w0 ? w0 : yr * ratio
         end
+
+        xm2 = set[:cols] * xr * set[:rows] * yr * set[:m]
+        next if xm2.round(2) == am2.round(2)
+
+        set[:dY] += (set[:d] - yr) / 2
+        set[:dX] += (set[:w] - xr) / 2 if set.key?(:dX)
+        set[:w ]  = xr
+        set[:d ]  = yr
+        set[:w0]  = set[:w] + gap
+        set[:d0]  = set[:d] + gap
+
+        skm2 -= (am2 - xm2)
+      end
+    end
+
+    # Size contraction: round 2: prioritize larger sets.
```

Comment:
Multiple rounds of skylight area contractions (instead of a single one), towards strictly meeting a requested **:srr** or **:area**.